diff -ruN proton-9.0/dlls/ntdll/unix/virtual.c proton-9.0-new/dlls/ntdll/unix/virtual.c
--- proton-9.0/dlls/ntdll/unix/virtual.c	2024-06-12 21:33:09.217784470 +0000
+++ proton-9.0-new/dlls/ntdll/unix/virtual.c	2024-06-12 21:34:14.103729743 +0000
@@ -456,41 +456,7 @@
 
 static void kernel_writewatch_register_range( struct file_view *view, void *base, size_t size )
 {
-    struct uffdio_register uffdio_register;
-    struct uffdio_writeprotect wp;
-
-    if (!(view->protect & VPROT_WRITEWATCH) || !use_kernel_writewatch) return;
-
-    madvise( base, size, MADV_NOHUGEPAGE );
-    if (use_kernel_writewatch == 2)
-    {
-        kernel_writewatch_reset( base, size );
-        return;
-    }
-
-    uffdio_register.range.start = (UINT_PTR)base;
-    uffdio_register.range.len = size;
-    uffdio_register.mode = UFFDIO_REGISTER_MODE_WP;
-    if (ioctl( uffd_fd, UFFDIO_REGISTER, &uffdio_register ) == -1)
-    {
-        ERR( "ioctl( UFFDIO_REGISTER ) failed, %s.\n", strerror(errno) );
-        return;
-    }
-
-    if (!(uffdio_register.ioctls & UFFDIO_WRITEPROTECT))
-    {
-        ERR( "uffdio_register.ioctls %s.\n", wine_dbgstr_longlong(uffdio_register.ioctls) );
-        return;
-    }
-    wp.range.start = (UINT_PTR)base;
-    wp.range.len = size;
-    wp.mode = UFFDIO_WRITEPROTECT_MODE_WP;
-
-    if (ioctl(uffd_fd, UFFDIO_WRITEPROTECT, &wp) == -1)
-    {
-        perror("ioctl(UFFDIO_WRITEPROTECT)");
-        exit(-1);
-    }
+	return;
 }
 
 static NTSTATUS kernel_soft_dirty_get_write_watches( void *base, SIZE_T size, void **addresses, ULONG_PTR *count, BOOL reset )
diff -ruN proton-9.0/dlls/winegstreamer/Makefile.in proton-9.0-new/dlls/winegstreamer/Makefile.in
--- proton-9.0/dlls/winegstreamer/Makefile.in	2024-06-12 21:33:09.449784276 +0000
+++ proton-9.0-new/dlls/winegstreamer/Makefile.in	2024-06-12 21:34:47.525701339 +0000
@@ -30,10 +30,4 @@
 	winegstreamer_classes.idl \
 	wm_reader.c \
 	wma_decoder.c \
-	wmv_decoder.c \
-	media-converter/audioconv.c \
-	media-converter/audioconvbin.c \
-	media-converter/fossilize.c \
-	media-converter/lib.c \
-	media-converter/murmur3.c \
-	media-converter/videoconv.c
+	wmv_decoder.c
diff -ruN proton-9.0/dlls/winegstreamer/unixlib.c proton-9.0-new/dlls/winegstreamer/unixlib.c
--- proton-9.0/dlls/winegstreamer/unixlib.c	2024-06-12 21:33:09.451784274 +0000
+++ proton-9.0-new/dlls/winegstreamer/unixlib.c	2024-06-12 21:35:39.125657224 +0000
@@ -46,8 +46,6 @@
 
 GST_DEBUG_CATEGORY(wine);
 
-extern bool media_converter_init(void);
-
 GstGLDisplay *gl_display;
 
 GstStreamType stream_type_from_caps(GstCaps *caps)
@@ -325,12 +323,5 @@
         }
     }
 
-    if (!media_converter_init())
-    {
-        GST_ERROR("Failed to init media converter.");
-        gst_object_unref(gl_display);
-        return STATUS_UNSUCCESSFUL;
-    }
-
     return STATUS_SUCCESS;
 }
